File: python/463.py
"""
Math problem
for each discontinuous 1s, we count 2 units for the perimeter
"""
class Solution(object):
    def islandPerimeter(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        count = 0
        xmax, ymax = len(grid), len(grid[0])
        for x in range(xmax):
            for y in range(ymax):
                if grid[x][y] and (y == 0 or not grid[x][y-1]):
                    count += 1
                if grid[x][y] and (x == 0 or not grid[x-1][y]):
                    count += 1
        return count * 2

File: solution/go/463.go
/*
For each "1", check its neighbors and update perimeter accordingly.
*/
func islandPerimeter(grid [][]int) int {
    perimeter := 0
    delta := [][]int{{-1, 0}, {1, 0}, {0, -1}, {0, 1}}
    for i := 0; i < len(grid); i++ {
        for j := 0; j < len(grid[0]); j++ {
            if grid[i][j] == 1 {
                perimeter += 4
                for _, d := range delta {
                    x, y := i + d[0], j + d[1]
                    if 0 <= x && x < len(grid) && 0 <= y && y < len(grid[0]) {
                        perimeter -= grid[x][y]
                    }
                }
            }
        }
    }
    return perimeter
}
